packages/runtime/container-runtime/src/summaryManagerUtils.ts

```diff
@@ -0,0 +1,200 @@
+        const isSummarizer = client.client.details?.type === summarizerClientType;
+        const newClient: IFullClient = {
+            clientId,
+            sequenceNumber: client.sequenceNumber,
+            isSummarizer,
+            prevClient: undefined,
+            nextClient: undefined,
+        };
+        this.clientMap.set(clientId, newClient);
+
+        if (newClient.isSummarizer) {
+            this.summarizerCount++;
+            this.emit("summarizerChange", this.summarizerCount);
+            return;
+        }
+
+        // Normal case is adding the latest client, which will bypass loop.
+        // Find where it belongs otherwise (this shouldn't happen, assert?).
+        assert(newClient.sequenceNumber > -1, "Negative sequence number not allowed");
+        let currClient = this.lastClient;
+        while (currClient.sequenceNumber > newClient.sequenceNumber) {
+            assert(currClient.prevClient !== undefined, "Previous client should always be defined");
+            // what to do if currClient === this.currentClient
+            currClient = currClient.prevClient;
+        }
+        // Now currClient is the node right before where the new client node should be.
+        newClient.prevClient = currClient;
+        newClient.nextClient = currClient.nextClient;
+        newClient.prevClient.nextClient = newClient;
+        if (newClient.nextClient === undefined) {
+            this.lastClient = newClient;
+        } else {
+            newClient.nextClient.prevClient = newClient;
+        }
+
+        if (this.currentClient === undefined && newClient.nextClient === undefined) {
+            this.currentClient = newClient;
+            this.emit("currentChange", this.getCurrentClient());
+        }
+        this.nonSummarizerCount++;
+    };
+
+    private readonly removeClient = (clientId: string) => {
+        const removeClient = this.clientMap.get(clientId);
+        if (removeClient !== undefined) {
+            this.clientMap.delete(clientId);
+            if (removeClient.isSummarizer) {
+                this.summarizerCount--;
+                this.emit("summarizerChange", this.summarizerCount);
+                return;
+            }
+
+            assert(removeClient.prevClient !== undefined, "Always should have prevClient");
+            removeClient.prevClient.nextClient = removeClient.nextClient;
+            if (removeClient.nextClient === undefined) {
+                this.lastClient = removeClient.prevClient;
+            } else {
+                removeClient.nextClient.prevClient = removeClient.prevClient;
+            }
+
+            if (removeClient === this.currentClient) {
+                this.currentClient = this.currentClient.nextClient;
+                this.emit("currentChange", this.getCurrentClient());
+            }
+            this.nonSummarizerCount--;
+        }
+    };
+
+    public getCurrentClient(): ITrackedClient | undefined {
+        return this.currentClient;
+    }
```

Comment:
could this be made simpler if this always just returne root.next, and increment was changed to remove root.next. what's the use of keep the invalid clients around

Reply:
We have the reset functionality, which we will likely need when all clients have been marked as bad summarizers we need to start over and probably just blow up or delay cycling through them again or something.